src/NativeScript/inlineFunctions.js

```diff
@@ -14,6 +14,103 @@ Object.assign(global, {
     NSMakeRange(location, length) {
         return new NSRange({ location, length });
     },
+
+    // Decorators support    
+    __decorate(decorators, target, key, desc) {
+        var c = arguments.length;
+        var r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc;
+        var d;
+
+        if (typeof global.Reflect === "object" && typeof global.Reflect.decorate === "function") {
+            r = global.Reflect.decorate(decorators, target, key, desc);
+        }
+        else {
+            for (var i = decorators.length - 1; i >= 0; i--) {
+                if (d = decorators[i]) {
+                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
+                }
+            }
+        }
+        return c > 3 && r && Object.defineProperty(target, key, r), r;
+    },
+    __param(paramIndex, decorator) {
+        return function (target, key) {
+            decorator(target, key, paramIndex);
+        }
+    },  
+
+    // Decorators  
+    ObjC(name) {
+        var additionalParams = [];
+        var hasName = (name !== undefined && typeof name === "string");
+        var argsShift = hasName ? 1 : 0;
+        for (var i = argsShift; i < arguments.length; i++) {
+            additionalParams[i - argsShift] = arguments[i];
+        }
+
+        return function (target, propertyKey, descriptor) {
+            if (propertyKey === undefined) { // Class decorator
+                var implementation = {};
+                var baseTarget = Object.getPrototypeOf(target);
+                
+                for (var prop in target.prototype) {
+                    if (target.prototype.hasOwnProperty(prop)) {
+                        implementation[prop] = target.prototype[prop];
+                    }
+                }
+
+                var newName = hasName ? name : target.name;                
+                var extended = baseTarget.extend(implementation, {
```

Comment:
> But IMO the ObjCMethod one should just accept the things concerning the method return type/name, and not accept a JSON with the parameters and everything. Also this creates some confusion to the end user. They will need to know if the param decorator or the method signature takes precedence.

Makes sense. I agree.

 > The benefit of using the baseTarget.extend is that we can specify the name of the class, which with the TS extends method is impossible.

We are setting the name of the native class to be the same as the name of the TS class and you can name your TS class however you want, so you have full control over the name of the native class. Still creating two classes when only one is needed is not only unnecessary but it possibly has a impact on the performance. However, if we really want to specify the name of the native class via a decorator (if we want the name not to be the same as the TS class), I suggest changing our TS extend to support it instead of calling `extend` twice.

P.S. Just want to add that it will be good if the decorator doesn't override the the `ObjCProtocols` static property but extend it. This way you can set some of your protocols  as a static property and some of them in a decorator.